### DISCUSSION TITLE
Dynamically-calculated eglot-workspace-configuration for python poetry venv's, what am I doing wrong?

### DISCUSSION THREAD

**Discussion by tomalexander** [joaotavora/eglot#967] [Q&A] [answered]
Hey, I am attempting to use eglot for python with pyright for the language server and poetry as the package manager. One of the nice aspects about poetry is every project gets its own dynamically-generated venv. They will be stored at a location like:
```
$HOME/.cache/pypoetry/virtualenvs/projname-szt0v1P8-py3.10
```
which means that my virtual environment path can change any time I update my python version or regenerate my venv. This makes setting `eglot-workspace-configuration` via `.dir-locals.el` impractical.

Instead, I've been trying to set up my config to discover the venv by running `poetry env info -p` and setting my `eglot-workspace-configuration` with that. Unfortunately though, I've been running into an issue where this works great for the first python file I open, but if I open a 2nd python file from another project, then it will re-use the venv from the first project.

I wrote a minimal example that doesn't depend on poetry to illustrate my point. This script below will:

1. Create a temporary directory
2. Create 2 venvs, one with `requests` installed and one with `pyyaml` installed
3. Create 2 git repos with a python file inside, one importing `requests` and the other importing `yaml`
4. Launch emacs with a minimal config to install eglot and set the `eglot-workspace-configuration` dynamically. For this minimal example, I just do a simple grep on the folder name to keep things simple.
5. Before the script exits, it will delete the temporary directory containing both projects and venvs.

```bash
#!/usr/bin/env bash
#
# Set up two separate projects with different git dependencies and
# separate eglot configs for pyright.
set -euo pipefail
IFS=$'\n\t'
DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" && pwd )"

WORK_DIR=$(mktemp -d -t eglot_test.XXXXXX)

function cleanup {
    rm -rf "$WORK_DIR"
}
for sig in EXIT INT QUIT HUP TERM; do
  trap "set +e; cleanup" "$sig"
done

FOO_VENV="$WORK_DIR/venv_foo"
BAR_VENV="$WORK_DIR/venv_bar"

cd "$WORK_DIR"

# set up "foo" project
python -m venv venv_foo
$FOO_VENV/bin/pip install requests
mkdir foo
git -C foo init
cat > foo/main.py <<EOF
import requests

print("foo")
EOF

# set up "bar" project
python -m venv venv_bar
$BAR_VENV/bin/pip install pyyaml
mkdir bar
git -C bar init
cat > bar/main.py <<EOF
import yaml

print("bar")
EOF

EMACS_INIT=$(cat <<EOF
(progn
  (package-initialize)
  (add-to-list 'package-archives
               '("melpa" . "https://melpa.org/packages/")
               )
  (when (not package-archive-contents) (package-refresh-contents))
  (unless (package-installed-p 'eglot) (package-install 'eglot))
  (add-hook 'python-mode-hook
            (lambda ()
              (when (string-match-p "foo" default-directory)
                (let ((venv (concat default-directory "/../venv_foo")))
                  (setq eglot-workspace-configuration (list (cons ':python (list ':venvPath venv ':pythonPath (concat venv "/bin/python")))))
                  (message "set the eglot venv to %s" venv)
                  )
                )
              (when (string-match-p "bar" default-directory)
                (let ((venv (concat default-directory "/../venv_bar")))
                  (setq eglot-workspace-configuration (list (cons ':python (list ':venvPath venv ':pythonPath (concat venv "/bin/python")))))
                  (message "set the eglot venv to %s" venv)
                  )
                )
              (eglot-ensure)
              ))
  )
EOF
)
emacs -q --eval "$EMACS_INIT" foo/main.py
```

If I launch the script, then it loads up `foo/main.py` and correctly identifies the venv so there are no flymake errors when I import `requests` (though it is greyed out because I don't actually use the dependency in the script)

![screenshot_2022-05-29-153636](https://user-images.githubusercontent.com/458272/170888495-d23007d2-fee3-4e66-9b6a-1138039ef2de.png)

If I switch to the `*EGLOT (foo/python-mode) events*` then I can see the venv path is being sent appropriately:
```elisp
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
	  (:settings
	   (:python
	    (:venvPath "/tmp/eglot_test.PwteD1/foo//../venv_foo" :pythonPath "/tmp/eglot_test.PwteD1/foo//../venv_foo/bin/python"))))
```

but if I `C-x C-f` and open `bar/main.py` it will show an error importing `yaml`:

![screenshot_2022-05-29-153955](https://user-images.githubusercontent.com/458272/170888581-6cbb1a34-0c51-45f8-bf21-f123f5f9cdb4.png)

and if I open `*EGLOT (bar/python-mode) events*` then I can see that eglot is still sending the venv from `foo`:
```elisp
(:jsonrpc "2.0" :method "workspace/didChangeConfiguration" :params
	  (:settings
	   (:python
	    (:venvPath "/tmp/eglot_test.PwteD1/foo//../venv_foo" :pythonPath "/tmp/eglot_test.PwteD1/foo//../venv_foo/bin/python"))))
```

Now to prove that its not just `bar` being incorrectly set up, if you close emacs and edit the last line of the above script so that its opening the `bar` project first:
```bash
emacs -q --eval "$EMACS_INIT" bar/main.py
```

Then you will see the exact opposite behavior, where `yaml` is loaded correctly and the venv path for bar's venv is sent properly but `C-x C-f` to open `foo/main.py` results in `import requests` having an error and eglot still sending the `bar` venv path.

I've run the same test using `.dir-locals.el` to set separate venvs and eglot worked perfectly with that, so I know it is possible to have eglot send different configs for different python buffers, but for some reason my python-mode hook isn't accomplishing that goal. Does anyone know what I'm doing wrong?

Finally, for completeness and to provide full context, here is the relevant bits on my `.emacs` for what I actually am using this for, with invoking poetry to get the venv path:
```elisp
(defun run-command-in-directory (dir cmd &rest args)
  "Run a command in the specified directory. If the directory is nil, the directory of the file is used. The stdout result is trimmed of whitespace and returned."
  (let (
        (default-directory (or dir default-directory))
        (stdout-buffer (generate-new-buffer "tmp-stdout" t))
        (full-cmd (append '(call-process cmd nil (list stdout-buffer nil) nil) args))
        )
    (unwind-protect
        (let ((exit-status (condition-case nil (eval full-cmd) (file-missing nil))))
          (if (eq exit-status 0)
              (progn
                (with-current-buffer stdout-buffer
                  (string-trim (buffer-string))
                  )
                )
            )
          )
      (kill-buffer stdout-buffer)
      )
    )
  )

(defun locate-venv-poetry ()
  "Find a poetry venv."
  (run-command-in-directory nil "poetry" "env" "info" "-p")
  )

(use-package python
  :mode ("\\.py\\'" . python-mode)
  :commands python-mode
  :pin manual
  :hook (
         (python-mode . (lambda ()
                          (when (executable-find "poetry")
                            (let ((venv (locate-venv-poetry))) (when venv
                                                                 (setq eglot-workspace-configuration
                                                                       (list (cons ':python (list ':venvPath venv ':pythonPath (concat venv "/bin/python")))))
                                                                 ))
                            )
                          (eglot-ensure)

                          ))
         )
  )
```

Possible relevant discussions:

1. This discussion seems to be asking for exactly what I need, but the author's needs were able to be met in a simpler manner with relative paths which does not work for poetry: https://github.com/joaotavora/eglot/discussions/878
2. This discussion might be pointing out a bug that might be causing the above behavior, but it may also be unrelated. Idk: https://github.com/joaotavora/eglot/discussions/876

**Comment by jdtsmith** [comment, marked as answer]
I can confirm this behavior.  The reason is subtle: the method `eglot-handle-request` for `workspace/configuration` uses `with-temp-buffer`.  But `eglot-workspace-configuration` is marked as a buffer-local variable.  So setting it as you are doing affects the buffer-local value in the python buffer of interest.  For the _first_ such buffer, its value becomes the _default_ value of `eglot-workspace-configuration`.  This same default value is active in all other buffers, including the temp-buffer from which `eglot-handle-request` responds to the server's config request.   So it works as you'd expect once, but only once. 

Though I couldn't understand why, there must be a reason for the use of `with-temp-buffer`, since eglot goes to the trouble of enabling the major-mode, and adding in the dir-local variables in it.  This is likely also why .dir-locals is the usual recommendation for eglot config.

A simple, cheap and rather unsatisfying workaround is to use `setq-default` instead of `setq`.

**Reply by tomalexander** [reply]
Ah thanks for figuring out the cause! I was worried `setq-default` might somehow impact the other eglot sessions since it is no longer setting a buffer-local variable at that point but in my very basic test of opening `foo/main.py` then `bar/main.py` then creating new files `foo/other.py` and `bar/other.py`, those new files both correctly had the right venv activated.

It definitely feels dirty to be using `setq-default` but I haven't noticed any downsides yet, so thank you!

**Reply by jdtsmith** [reply]
It's pretty dirty.  And if the server asks _again_ for the configuration, and you have in the meantime started yet another server, it will provide incorrect info.  

I believe @joaotavora is thinking about a replacement variable that could optionally be set to a callback function to provide config details based on the path.  He makes the good point that some servers can be running without any attached Emacs buffers, and still need to know their config; hence the heavy reliance on dir-locals via `with-temp-buffer`.

**Reply by joaotavora** [reply]
@tomalexander you're of course entitled to your opinion, but IMO @jdtsmith had a misunderstanding regarding the buffer-local nature of the variable in question.  Hopefully that is cleared up.  That "buffer-local" nature is simply an implementation detail, which shouldn't concern the user.  

One should  think of the variable as "directory-local", because that's closest to its intended purpose.  In Eglot, a project's root directory _is_ the workspace.

_**Now, how to set it?**_  One way to set it is in a `.dir-locals.el`in the root directory of the workspace, of course.  But there are other means of setting directory-local variables via Elisp (check the manual: 49.2.5 Per-Directory Local Variables) which you may find more suitable if adding a file to your repo isn't a good option.

_**What to set it to?**_ That depends on your server and the JSON object it wants to receive when it requests the value.

_**But wait, there's this TRAMP bug!**_ Yes that's true, there's a bug with TRAMP where the server doesn't provide enough information for Eglot to guess where is the directory (this is #876).  The bug has been observed by some people, but I've yet to receive a reliable reproduction recipe. But the variable should still be dir-local.

_**But wait, I have a Python system with a pretentious name that needs a dynamically calculated value!**_.  You need the value to vary over time, depending on dynamic context.  You need a function. This particular variable in Eglot doesn't have that yet, so you need this patch :point_down:  . I hope you or others can test so I can merge it to Eglot.  But remember,you still set the variable dir-locally.  If you have problems applying the patch, let me know I'll do a branch.

```diff
diff --git a/eglot.el b/eglot.el
index 0b64cd2..9bb1dcd 100644
--- a/eglot.el
+++ b/eglot.el
@@ -2199,8 +2199,13 @@ Records BEG, END and PRE-CHANGE-LENGTH locally."
 
 (defvar-local eglot-workspace-configuration ()
   "Alist of (SECTION . VALUE) entries configuring the LSP server.
-SECTION should be a keyword or a string, value can be anything
-that can be converted to JSON.")
+SECTION should be a keyword or a string.  VALUE is a
+plist or a primitive type converted to JSON.
+
+The value of this variable can also be a unary function of a
+`eglot-lsp-server' instance, the server connection requesting the
+configuration.  It should return an alist of the format described
+above.")
 
 ;;;###autoload
 (put 'eglot-workspace-configuration 'safe-local-variable 'listp)
@@ -2209,16 +2214,19 @@ that can be converted to JSON.")
   "Send a `:workspace/didChangeConfiguration' signal to SERVER.
 When called interactively, use the currently active server"
   (interactive (list (eglot--current-server-or-lose)))
-  (jsonrpc-notify
-   server :workspace/didChangeConfiguration
-   (list
-    :settings
-    (or (cl-loop for (section . v) in eglot-workspace-configuration
-                 collect (if (keywordp section)
-                             section
-                           (intern (format ":%s" section)))
-                 collect v)
-        eglot--{}))))
+  (let ((config (if (functionp eglot-workspace-configuration)
+                    (funcall eglot-workspace-configuration server)
+                  eglot-workspace-configuration)))
+    (jsonrpc-notify
+     server :workspace/didChangeConfiguration
+     (list
+      :settings
+      (or (cl-loop for (section . v) in config
+                   collect (if (keywordp section)
+                               section
+                             (intern (format ":%s" section)))
+                   collect v)
+          eglot--{})))))
 
 (cl-defmethod eglot-handle-request
   (server (_method (eql workspace/configuration)) &key items)
```

**Reply by jdtsmith** [reply]
> Hopefully that is cleared up. That "buffer-local" nature is simply an implementation detail, which shouldn't concern the user.

Yes, clear now, as pointed out above.

> you need this patch

Trying now.  Is `default-directory` in `eglot-signal-didChangeConfiguration` always within the workspace?  Also, I noticed another use of `eglot-workspace-configuration ` in `eglot-handle-request` (where default-directory is set to the scopeUri).

**Reply by joaotavora** [reply]
I've pushed a branch `fix-967-eglot-workspace-configuration-can-be-a-function` containing the patch and also fleshing out the README.md section about this variable a  bit more.  Fell free to add examples to the README.md about your usage of `eglot-workspace-configuration`.

**Reply by joaotavora** [reply]
> Is default-directory in eglot-signal-didChangeConfiguration always within the workspace? 

Within the function that you set `eglot-workspace-configuration` to, `default-directory` should be the root of the project (aka lsp workspace).  Please confirm that it is/isn't.

> Also, I noticed another use of eglot-workspace-configuration in eglot-handle-request (where default-directory is set to the scopeUri).

Thanks. There was a bug in the patch.  Please check out the branch again.

**Reply by jdtsmith** [reply]
Thanks for the branch.  In trying it I notice that that second scopeUri-based usage of workspace config does seem essential, as otherwise I get an error on the `workspace/configuration` request and no config is actually set (since apparently my server uses the "PULL" method of asking for config):

```
(:jsonrpc "2.0" :id 2 :method "workspace/configuration" :params
	  (:items
	   [(:scopeUri "file:///Users/jdsmith/code/python/pahfit" :section "python")]))
[client-reply] (id:2) ERROR Mon Jul 18 16:56:31 2022:
(:jsonrpc "2.0" :id 2 :error
	  (:code -32603 :message "Internal error"))
[server-notification] Mon Jul 18 16:56:32 2022:
(:jsonrpc "2.0" :method "window/logMessage" :params
	  (:type 1 :message "Error reading settings: Error: Internal error"))

```

I needed a similar small patch to handle-request, then it works great.  

```diff
@@ -2240,10 +2240,13 @@ When called interactively, use the currently active server"
                        (if (and (not (string-empty-p uri-path))
                                 (file-directory-p uri-path))
                            (file-name-as-directory uri-path)
-                         (project-root (eglot--project server)))))
+                         (project-root (eglot--project server))))
+                     (config (if (functionp eglot-workspace-configuration)
+                                 (funcall eglot-workspace-configuration server)
+                               eglot-workspace-configuration)))
                 (setq-local major-mode (eglot--major-mode server))
                 (hack-dir-local-variables-non-file-buffer)
-                (alist-get section eglot-workspace-configuration
+                (alist-get section config
                            nil nil
                            (lambda (wsection section)
                              (string=
```

**Reply by jdtsmith** [reply]
Great; with your latest fix this works well, using a custom function (very germane for this discussion):

```elisp
(defun my/eglot-workspace-config (server)
          (if-let ((venv (simple-venv-find))
		   ((not (eq venv 'none))))
	      (list (cons :python
			  (list :venvPath venv
				:pythonPath (simple-venv-interpreter venv))))))
```

(N.B. I have not yet released `simple-venv`, but will).  Is there a way to query the `server` object regarding which mode it is servicing and command it is running (pyright vs. pylsp, for example)?  Then I could use that information to return various workspace config for different servers.

> Within the function that you set eglot-workspace-configuration to, default-directory should be the root of the project 

This does seem to be the case.  Thanks again.

**Reply by joaotavora** [reply]
> Is there a way to query the server object 

Yes, but only with internal accessors, such as `eglot--major-mode`.  Tell me which ones you need and we can think about making some of them public.  However, if you set the variable dir-locally according to the instructions above (which I have fleshed out in the README.md, improved in the latest branch) you shouldn't need to do much querying for major modes.

**Reply by jdtsmith** [reply]
Since the custom function could be called w.r.t. any server, at least the server name seems important, so the user function can tailor its response.  

I had not realized you could update dir-local vars without the actual file.  But I suspect the custom user function is more flexible, for example if you are running two different servers for the same project directory for code from more than one language.

**Reply by joaotavora** [reply]
> I had not realized you could update dir-local vars without the actual file. But I suspect the custom user function is more flexible,

They serve different purposes, so I don't understand the comparison:

1. The first facility, `dir-locals-set-class-variables`, is for not having to write an actual `.dir-locals.el` into the repository, or simply for managing the contents of that would-be file from Elisp code located elsewhere.  I use it a lot for storing per-repo settings in big shared repos where noone uses Emacs (fools!)

2. The second facility is for having a run-time, dynamic decision based on `default-directory`, the time of day, color of sky, etc as to what the "workspace configuration" should be.

> Since the custom function could be called w.r.t. any server, at least the server name seems important, so the user function can tailor its response.

Can you give a short example of what that user function look like if I gave you that accessor under the name `eglot-server-name`?

**Comment by jdtsmith** [comment]
> They serve different purposes

What I meant is that you can `(eval  . forms)` in dir-locals to eval arbitrary code, so presumably could update the psuedo-dir-locals vars for a given workspace directory whenever you wanted.  In that sense (I think) it would function somewhat like a more convoluted custom user function.

> Can you give a short example of what that user function look like if I gave you that accessor under the name eglot-server-name?

Something like:
```elisp
((defun my/eglot-workspace-config (server)
  (let ((mode (eglot-server-mode server))
	(server (eglot-server-name server)))
    (cond
     ((equal mode "python")
      (cond
       ((equal server "pyright")
	(list (cons :python
		    (list :venvPath (my/path default-directory)
			  :pythonPath some-interpreter))))
       ((equal server "pylsp")
	(list ...))))
     ((equal mode "javascript")
      (list ...))
     ((equal mode "go")
      (list ...)))))
```

**Comment by joaotavora** [comment]
> What I meant is that you can (eval . forms) in dir-locals to eval arbitrary code, so presumably could update the psuedo-dir-locals vars for a given workspace directory whenever you wanted. In that sense (I think) it would function somewhat like a more convoluted custom user function.

Right, it could maybe help you here, but it's _still_ not the same, as that evaluation happens once when you visit the buffer.  So it's not the same as a function called when the server wants to know the value of something.  Not to mention it's very hard to debug and generally not a good thing.

> Something like:

Right, as i thought, you just need to return an alist with all possible _section names_ (`:python`, `:pyls`, `:gols`) as the keys of the alist return value.  Your function will become simpler and just as effective. Alternatively, craft different functions for different major modes and set the dir-local value of `eglot-workspace-configuration` to  different functions for each major mode.

This is not only simpler, but safer, as the "server name" you're possibly looking for is not easy to define, and may not be portable across servers (if I happen to be mistaken, please point me to the LSP spec on server names).

**Comment by jdtsmith** [comment]
> not to mention it's very hard to debug and generally not a good thing.

Agreed.  

> return an alist with all possible section names

Good thought thanks.

> This is not only simpler, but safer, as the "server name" you're possibly looking for is not easy to define, and may not be portable across servers

The issue I posted earlier points out that the workspace/config "section" names and their contents are underspecified, and a bit of a mishmash across various servers.   But no sense solving a problem that hasn't yet appeared (for me anyway).

**Comment by rafonseca** [comment]
I know this is not exactly what was asked here, but I would like to point out a simple solution for the overall problem of using the right env with pyright and poetry:

```
(with-eval-after-load 'eglot
  (add-to-list 'eglot-server-programs
               `((python-mode python-ts-mode) .
		 ,(eglot-alternatives '("pylsp" "pyls" ("poetry" "run" "pyright-langserver" "--stdio")  ("pyright-langserver" "--stdio") "jedi-language-server")))))
```

Launching the server with "poetry run" let poetry set up the environment for us. There is a minor issue on alternatives prompt if one uses the same  strategy for other LSPs.

**Reply by joaotavora** [reply]
`poetry run` seems to be a great idea to solve this, much more elegant. What is the "issue" you are facing? If you only use poetry for your python projects, you probably don't need `eglot-alternatives`. Or you can even set a dir-local value for `eglot-server-commands`...

**Reply by rafonseca** [reply]
It is actually an edge case. If I want to switch between jedy and pyright using "poetry run ..." for both, the alternatives prompt shows two "poetry" entries.  There are many workarounds for this. I dont think this issue should be addressed.
